### PR TITLE
Make more lenient CIS-2 Lookup

### PR DESCRIPTION
**Pull request by ramakser**
## Purpose
- More lenient CIS-2 Lookup. 
- Fix CIS-2 token preview in case when thumbnail data is missed
- Fix token fetch logic for the smart contracts with large amount of tokens. Handle broken token metadata. 

## Changes

- Add paginated call for 'CIS2TokenSelectView'  to show list of the tokens. 
- Add new API's v1 for `v1/CIS2TokenMetadata` and `v1/CIS2TokenBalance`
- Fixed corrupted or missed token metadata handle logic.  Filter out broken tokens from the list
- Fixed thumbnails for CIS-2 in case when thumbnail is no available in metadata  - use display image instead. [Docs](https://proposals.concordium.software/CIS/cis-2.html#token-metadata-json)

### CODE REVIEW

File: ConcordiumWallet/Views/AccountsView/AccountDetails/TokenLookup/CIS2Service.swift [hunk 1]
```diff
@@ -100,17 +100,11 @@ class CIS2Service: CIS2ServiceProtocol {
                 url: ApiConstants.cis2Tokens
                     .appendingPathComponent(contractIndex)
                     .appendingPathComponent(contractSubindex),
-                parameters: ["limit" : "1000"]
+                parameters: ["limit" : "20"]
```
Comment (ramakser) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/TokenLookup/CIS2Service.swift:103]: we probably should add here a paginated call, due to current API v1 returns max 20 items in response

**Review by Radiokot — APPROVED** [review]

File: ConcordiumWallet/Views/AccountsView/AccountDetails/TokenSelectionView/CIS2TokenSelectViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                    if currentPage == 1 {
+                        tokens = representables
+                    } else {
+                        tokens += representables
+                    }
+                }
+            } catch {
+                await MainActor.run {
+                    isLoading = false
+                }
```
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/TokenSelectionView/CIS2TokenSelectViewModel.swift:88]: Consider doing this in a `defer` instead.

File: ConcordiumWallet/Views/AccountsView/AccountDetails/TokenSelectionView/CIS2TokenSelectViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            
+            for try await data in group {
+                allData.append(data)
+            }
+        }
+        
+        return allData.compactMap { (metadataItem, tokenMetadataDetails) -> (CIS2TokensMetadataItem, CIS2TokenMetadataDetails)? in
+            guard let tokenMetadataDetails = tokenMetadataDetails else { return nil }
```
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/TokenSelectionView/CIS2TokenSelectViewModel.swift:123]:
```suggestion
            guard let tokenMetadataDetails else { return nil }
```

Comment (ramakser) [reply ConcordiumWallet/Views/AccountsView/AccountDetails/TokenSelectionView/CIS2TokenSelectViewModel.swift:123]: I need this unwrapped value ` -> (CIS2TokensMetadataItem, CIS2TokenMetadataDetails)`. basically here we filter out  all tokens with empty or broken metadata, and return pair for rest

Comment (bisgardo) [reply ConcordiumWallet/Views/AccountsView/AccountDetails/TokenSelectionView/CIS2TokenSelectViewModel.swift:123]: The suggested change is just syntactic sugar, the semantics is the same.

**Pull request merged**